text/0000-projections-lifetimes-and-wf.md

```diff
@@ -0,0 +1,980 @@
+with type definitions, where the type is in charge of figuring out its
+own requirements and listing them in one place.)
+
+As the name "implied lifetime bounds" suggests, we currently limit
+implied bounds to region relationships. That is, we will implicitly
+derive a bound like `K:'a` or `V:'a`, but not `K:Hash` -- this must
+still be written manually. It might be a good idea to change this, but
+that would be the topic of a separate RFC.
+
+Currently, implied bound are limited to fn bodies. This RFC expands
+the use of implied bounds to cover impl definitions as well, since
+otherwise the annotation burden is quite painful. More on this in the
+next section.
+
+*NB.* There is an additional problem concerning the interaction of
+implied bounds and contravariance ([#25860]). To better separate the
+issues, this will be addressed in a follow-up RFC that should appear
+shortly.
+
+#### Missing WF checks
+
+Unfortunately, the compiler currently fails to enforce WF in several
+important cases. For example, the
+[following program](http://is.gd/6JXjyg) is accepted:
+
+```rust
+struct MyType<T:Copy> { t: T }
+
+trait ExampleTrait {
+    type Output;
+}
+
+struct ExampleType;
+
+impl ExampleTrait for ExampleType {
+    type Output = MyType<Box<i32>>;
+    //            ~~~~~~~~~~~~~~~~
+    //                   |
+    //   Note that `Box<i32>` is not `Copy`!
+}
+```
+
+However, if we simply naively add the requirement that associated
+types must be well-formed, this results in a large annotation burden
+(see e.g. [PR 25701](https://github.com/rust-lang/rust/pull/25701/)).
+For example, in practice, many iterator implementation break due to
+region relationships:
+
+```rust
+impl<'a, T> IntoIterator for &'a LinkedList<T> { 
+   type Item = &'a T;
+   ...
+}
+```
+
+The problem here is that for `&'a T` to be well-formed, `T: 'a` must
+hold, but that is not specified in the where clauses. This RFC
+proposes using implied bounds to address this concern -- specifically,
+every impl is permitted to assume that all types which appear in the
+impl header (trait reference) are well-formed, and in turn each "user"
+of an impl will validate this requirement whenever they project out of
+a trait reference (e.g., to do a method call, or normalize an
+associated type).
+
+# Detailed design
+
+This section dives into detail on the proposed type rules.
+
+### A little type grammar
+
+We extend the type grammar from [RFC 192] with projections and slice
+types:
+
+    T = scalar (i32, u32, ...)        // Boring stuff
+      | X                             // Type variable
+      | Id<P0..Pn>                    // Nominal type (struct, enum)
+      | &r T                          // Reference (mut doesn't matter here)
+      | O0..On+r                      // Object type
+      | [T]                           // Slice type
+      | for<r..> fn(T1..Tn) -> T0     // Function pointer
+      | <P0 as Trait<P1..Pn>>::Id     // Projection
```

Comment:
@pnkfelix 

> nit: should be <T0 as Trait<P1..Pn>>::Id, right?

yeah, I intentionally made it P0 so that I could cleanly say things like `\forall i. Pi` etc